File: coodingpenguin/programmers/level2/행렬테두리회전하기.py
dirs = [(0, 1), (1, 0), (0, -1), (-1, 0)]


def solution(rows, columns, queries):
    matrix = [[j * columns + i for i in range(1, columns + 1)] for j in range(rows)]
    result = []
    for y1, x1, y2, x2 in queries:
        min_num = 10000  # 최소 숫자
        cy, cx, cd = y1 - 1, x1 - 1, 0  # 현재 위치(y, x)와 이동 방향
        new_matrix = [row[:] for row in matrix]
        # 모든 방향을 돌 때까지 반복
        while cd < 4:
            min_num = min(min_num, matrix[cy][cx])  # 최솟값 갱신
            ny, nx = cy + dirs[cd][0], cx + dirs[cd][1]  # 이동할 위치
            # 범위에서 벗어나면
            if (ny < y1 - 1 or ny > y2 - 1) or (nx < x1 - 1 or nx > x2 - 1):
                cd += 1  # 방향을 바꾼다
                continue
            new_matrix[ny][nx] = matrix[cy][cx]
            cy, cx = ny, nx
        matrix = new_matrix
        result.append(min_num)
    return result

File: sally/programmers/week3/행렬 테두리 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <cmath>
#include <queue>
using namespace std;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    
    // init
    int arr[101][101] = {0,};
    int k = 0;
    for(int i = 0; i < rows; i++)
        for(int j = 0; j < columns; j++)
            arr[i][j] = ++k;

    int dir[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

    // roll
    int T = queries.size();
    for(int t = 0; t < T; t++){
        int x1 = queries[t][0]-1;
        int y1 = queries[t][1]-1;
        int x2 = queries[t][2]-1;
        int y2 = queries[t][3]-1;
        
        queue<pair<int, int>> loc;
        queue<int> val;
        
        vector<int> cnt;
        cnt.push_back(y2-y1);
        cnt.push_back(x2-x1);
        cnt.push_back(y2-y1);
        cnt.push_back(x2-x1);
        
        int cur_x = x1;
        int cur_y = y1;
        int d = 0;
        for(int i = 0; i < 4; i++){
            for(int j = 0; j < cnt[i]; j++){
                loc.push({cur_x, cur_y});
                val.push(arr[cur_x][cur_y]);
                cur_x += dir[d][0];
                cur_y += dir[d][1];
            }
            d++;
        }
        
        pair<int, int> tmp = loc.front();
        loc.pop();
        loc.push(tmp);
        
        int min_val = 1000001;
        while(!loc.empty()){
            arr[loc.front().first][loc.front().second] = val.front();
            min_val = min(min_val, val.front());
            loc.pop();
            val.pop();
        }
        answer.push_back(min_val);
    }
    
    return answer;
}